drivers/saul/init_devs/auto_init_saul_pwm.c

```diff
@@ -0,0 +1,179 @@
+/* The static variable will be unused in the 0 case and thus not emitted. */
+static saul_reg_t saul_reg_entries_dimmer[MAX(SAUL_PWM_DIMMER_NUMOF, 1)];
+
+/**
+ * @brief   Reference to the driver for single-channel dimmers
+ */
+extern saul_driver_t dimmer_saul_driver;
+
+/**
+ * @brief   Reference to the driver for RGB LEDs
+ */
+extern saul_driver_t rgb_saul_driver;
+
+/**
+ * Configure a PWM driver for LED output (1kHz, 8bit)
+ */
+static int configure(pwm_t dev)
+{
+    LOG_DEBUG("[auto_init_saul] initializing PWM %u for LED operation,", dev);
+    uint32_t freq = pwm_init(dev, PWM_LEFT, SAUL_PWM_FREQ, saul_pwm_resolution);
+    LOG_DEBUG(" actual frequency %lu,\n", freq);
+    return freq != 0 ? 0 : -ENOTSUP;
+}
+
+/**
+ * Configure the PWM driver at the given index (inside saul_pwm_dimmer_params,
+ * overflowing into saul_pwm_rgb_params) unless that device came up previously,
+ * in which case the function returns without any action.
+ * */
+static int configure_on_first_use(pwm_t dev, int index)
+{
+    for (int i = 0; i < (int)SAUL_PWM_DIMMER_NUMOF; ++i) {
+        pwm_t currentdev = saul_pwm_dimmer_params[i].channel.dev;
+        if (currentdev == dev) {
+            if (i == index) {
+                return configure(dev);
+            }
+            return 0;
+        }
+    }
+    for (int i = 0; i < (int)SAUL_PWM_RGB_NUMOF; ++i) {
+        for (int j = 0; j < 3; ++j) {
+            int index = SAUL_PWM_DIMMER_NUMOF + i * 3 + j;
+            pwm_t currentdev = saul_pwm_rgb_params[i].channels[j].dev;
+            if (currentdev == dev) {
+                if (i == index) {
+                    return configure(dev);
+                }
+                return 0;
+            }
+        }
+    }
+    return -ENOENT;
+}
+
+void auto_init_pwm(void)
+{
+    for (int i = 0; i < (int)SAUL_PWM_DIMMER_NUMOF; i++) {
+        const saul_pwm_dimmer_params_t *p = &saul_pwm_dimmer_params[i];
+
+        LOG_DEBUG("[auto_init_saul] initializing dimmer #%u\n", i);
+
+        saul_reg_entries_dimmer[i].dev = (void *)p;
+        saul_reg_entries_dimmer[i].name = p->name;
+        saul_reg_entries_dimmer[i].driver = &dimmer_saul_driver;
+
+        int err = configure_on_first_use(p->channel.dev, i);
+        if (err != 0) {
+            LOG_ERROR(
+                "[auto_init_saul] Error initializing device for dimmer #%u\n",
+                i);
+            /* not `continue`ing: we could run into this on a non-first use and
+             * then we couldn't break either */
+        }
+        /* set initial dark state */
+        phydat_t s;
+        s.val[0] = 0;
+        saul_reg_entries_dimmer[i].driver->write(p, &s);
+        /* add to registry */
+        saul_reg_add(&(saul_reg_entries_dimmer[i]));
+    }
+
+    for (int i = 0; i < (int)SAUL_PWM_RGB_NUMOF; i++) {
+        const saul_pwm_rgb_params_t *p = &saul_pwm_rgb_params[i];
+
+        LOG_DEBUG("[auto_init_saul] initializing RGB #%u\n", i);
+
+        saul_reg_entries_rgb[i].dev = (void *)p;
```

Comment:
```suggestion
        saul_reg_entries_rgb[i].dev = p;
```

Explicit casts disable the limited type safety C provides. Thus, it is better to avoid them whenever possible. From and to `void *` can always be don implicitly.

One example where this can shoot in you foot:

```C
    int *foo = (int *)malloc(4 * sizeof(int));
    int *bar = malloc(4 * sizeof(int));
```

If the `#include <stdlib.h>` was forgotten and implicit function declarations from the ancient past of C were not disabled via compiler flags, the compiler would assume `int malloc(size_t)` as signature for `malloc()`. The first version with the explicit cast would generate no warning, the one with the implicit cast would warn.

So, IMO we should avoid explicit casts unless absolutely needed.

Reply:
While the other group of casts was actually needless, this *is* required: as to discard the const around the pointer. (The promise never to write there is upheld later when -- now implicitly, as per the other comment -- casting back to a `const saul_pwm_dimmer_params_t *`).